### PR TITLE
 Release Planning and Roadmapping : Roadmap Visualization & Core Export #127

### PR DESCRIPTION
**Pull request by Gowrishbyndoor**
Issue: #127

## Objective

Implement API endpoints and data aggregation logic to support roadmap visualization, release dashboards, and comprehensive reporting.

### 1. Roadmap API

`GET /api/roadmap` - Get product roadmap with releases, features, progress metrics and health indicators.

#### Sorting:
Releases in the response should be sorted in descending order by date, 
that is taken from the first of the following fields containing a non-null value:
1. `actualReleaseDate`
2. `releasedAt`
3. `plannedReleaseDate`
4. `createdAt`

#### Filters:
Support filters as query parameters:
  - `productCodes` : `string[]` (optional) - list of product codes to filter releases by
  - `statuses` :  `string[]` (optional) - list of release statuses to filter releases by
    (e.g., `DRAFT`, `PLANNED`, `IN_PROGRESS` etc.)
  - `dateFrom` : iso date (optional) - filter releases by the same date field as used for sorting above (inclusive)
  - `dateTo` : iso date (optional)
  - `groupBy` : `productCode | status | owner` (optional) - group releases by the specified field, keep same sorting within each group. Groups themselves should be sorted by the date of the first release.
  - `owner` : string (optional) - return releases for specified owner only
    - return empty results if such user does not exist or doesn't own any releases

#### Response:
```
{
  "roadmapItems": [
    {
      "product": {
        "id": Long,
        "code": String
      },
      "release": {
        "id": Long,
        "code": String,
        "description": String,
        "status": String,
        "releasedAt": DateTime,
        "plannedStartDate": DateTime,
        "plannedReleaseDate": DateTime,
        "actualReleaseDate": DateTime,
        "owner": String,
        "notes": String,
        "createdBy": String,
        "createdAt": DateTime,
        "updatedBy": String,
        "updatedAt": DateTime
      },
      "progressMetrics": {
        "totalFeatures": Integer,
        "completedFeatures": Integer,
        "inProgressFeatures": Integer,
        "newFeatures": Integer,
        "onHoldFeatures": Integer,
        "completionPercentage": Double
      },
      "healthIndicators": {
        "riskLevel": String,
        "timelineAdherence": String
      },
      "features": [
        {
          "id": Long,
          "code": String,
          "title": String,
          "description": String,
          "status": String,
          "releaseCode": String,
          "assignedTo": String,
          "createdBy": String,
          "createdAt": DateTime,
          "updatedBy": String,
          "updatedAt": DateTime
        }
      ]
    }
  ],
  "summary": {
    "totalReleases": Integer,
    "completedReleases": Integer,
    "draftReleases": Integer,
    "totalFeatures": Integer,
    "overallCompletionPercentage": Double
  },
  "appliedFilters": {
    "productCodes": String[],
    "statuses": String[],
    "dateFrom": LocalDate,
    "dateTo": LocalDate,
    "groupBy": String,
    "owner": String
  }
}
```

##### Health Indicators:
- `riskLevel` is calculated based on percentage of features in the release in status `ON_HOLD`, and can be one of the following:
  - `ZERO`(0 features on hold)
  - `LOW` (<10% features on hold)
  - `MEDIUM` (>10%)
  - `HIGH` (>30%)
- `timelineAdherence` is calculated based on how many days are left from now until `plannedReleaseDate` for upcoming releases,
  or by comparing `actualReleaseDate` (with fallback to `releasedAt`) vs `plannedReleaseDate` for finished releases.
  It can be one of the following: 
  - null - if `plannedReleaseDate` is missing
  - `ON_SCHEDULE` - if either `actualReleaseDate` or `releasedAt` is set and is before or on the planned release date.
    - if both dates are missing, compare current date to planned release date.
  - `DELAYED` - if we are past the planned release date, but delay is less than 14 days
    - apply same field selection logic
  - `CRITICAL` - 14+ days delay

#### Validation:
Return 400 when:
- query parameters don't match expected filter types
  - ignore and allow extra query parameters in the request, that are not on the list of supported filters
- list of statuses contains incorrect status values (case-insensitive)
- `dateFrom` is after `dateTo`
- `groupBy` value is not on the list of supported values (case-insensitive)

### 2. Export API

`GET /api/roadmap/export?format=CSV|PDF` - Export roadmap as a file

- Support all the same filters and apply the same sorting as in the roadmap API
- `format` parameter is mandatory; allowed values are `CSV` and `PDF` (case-insensitive)
- Exported file name should be `Roadmap_{timestamp in yyyyMMddHHmmss format}.csv/pdf`
- CSV file should contain the following columns:
  - Product Code
  - Release Code
  - Release Description
  - Release Status
  - Released At
  - Planned Start Date
  - Planned Release Date
  - Actual Release Date
  - Owner
  - Total Features
  - Completed Features
  - In Progress Features
  - New Features
  - On Hold Features
  - Completion Percentage
  - Timeline Adherence
  - Risk Level
- PDF export should contain the same data in a readable tabular format

## Implementation
Service Layer:
- Implement `RoadmapService` for roadmap data aggregation and filtering
- Implement `ReportingService` for roadmap export functionality

## Test Coverage
- Integration tests for the roadmap endpoint, covering:
  - all filters and edge cases
  - health/progress metrics in roadmap responses
  - appliedFilters are present and accurate
  - validation of filters
- Integration tests for roadmap export endpoint for both CSV and PDF formats

## Acceptance Criteria
- Roadmap API returns accurate, filtered, and grouped data.
- All filters (productCode, dateRange, includeCompleted, groupBy) work correctly.
- Filter validations correctly reject malformed requests.
- Additional query parameters in the request are allowed, but ignored without error.
- Health and progress metrics are included in roadmap responses.
- Export functionality works for all supported formats.
- Response times meet performance requirements (<500ms for roadmap search, <2s for export)

FAIL_TO_PASS: RoadmapControllerIntegrationTests

### CODE REVIEW

File: src/main/java/com/sivalabs/ft/features/domain/mappers/ReleaseMapper.java [hunk 1]
```diff
@@ -4,7 +4,9 @@
 import com.sivalabs.ft.features.domain.entities.Release;
 import org.mapstruct.Mapper;
 
-@Mapper(componentModel = "spring")
+@Mapper(
+        componentModel = "spring",
+        uses = {ProductMapper.class})
```
[the remark this answers is not in the document]
Comment (Gowrishbyndoor) [reply src/main/java/com/sivalabs/ft/features/domain/mappers/ReleaseMapper.java]: Needed this block as product is needed.

File: src/test/java/com/sivalabs/ft/features/api/controllers/RoadmapControllerIntegrationTests.java [hunk 2]
```diff
@@ -0,0 +1,1326 @@
...
+
+            Map<String, Object> appliedFilters = (Map<String, Object>) response.get("appliedFilters");
+            assertThat(appliedFilters.get("dateFrom")).isEqualTo("2024-01-01");
+            assertThat(appliedFilters.get("dateTo")).isEqualTo("2025-12-31");
+
+            // Validate that all returned releases are within date range
+            for (Map<String, Object> item : roadmapItems) {
+                Map<String, Object> release = (Map<String, Object>) item.get("release");
+                String releaseDate = (String) release.get("releasedAt");
+                if (releaseDate != null) {
```
[the remark this answers is not in the document]
Comment (Gowrishbyndoor) [reply src/test/java/com/sivalabs/ft/features/api/controllers/RoadmapControllerIntegrationTests.java]: I have removed the if condition assertion where ever possible, in few cases it requires there so i have kept, example when test is about PlannedReleaseDateMissing then we need to put a if condition for plannedReleaseDate is null, in such scenario i have kept as it is. in remaining scenario i have removed if conditions.

**Comment by Gowrishbyndoor** [comment]
@issue-validator validate this PR

**Comment by Gowrishbyndoor** [comment]
Recent validation successful in https://github.com/dpaia/infrastructure/actions/runs/21077504034/job/60622990513

File: src/main/java/com/sivalabs/ft/features/domain/mappers/ReleaseMapper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gowrishbyndoor) [reply src/main/java/com/sivalabs/ft/features/domain/mappers/ReleaseMapper.java]: This one https://github.com/dpaia/feature-service/pull/170/files#:~:text=%40JsonIgnore%20ProductInfo%20product%29%20%7B%7D

**Comment by innokenty** [comment]
@issue-validator validate this PR
